File: backend-jobflow/src/main/java/com/jobflow/service/ApplicationService.java
package com.jobflow.service;

import com.jobflow.dto.request.ApplicationRequest;
import com.jobflow.dto.response.ApplicationResponse;
import com.jobflow.dto.response.ApplicationStatusHistoryResponse;
import com.jobflow.entity.ApplicationStatus;
import com.jobflow.entity.Priority;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface ApplicationService {
    ApplicationResponse create(ApplicationRequest request);
    ApplicationResponse update(UUID id, ApplicationRequest request);
    void delete(UUID id);
    ApplicationResponse get(UUID id);
    Page<ApplicationResponse> list(String search, ApplicationStatus status, Priority priority, UUID companyId,
                                    LocalDate dateFrom, LocalDate dateTo, Pageable pageable);
    ApplicationResponse updateStatus(UUID id, ApplicationStatus newStatus);
    List<ApplicationStatusHistoryResponse> getStatusHistory(UUID id);
}

File: frontend-jobflow/src/app/core/services/application.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { environment } from '../../../environments/environment';
import { Page } from '../models/company.model';
import {
  Application,
  ApplicationRequest,
  ApplicationStatus,
  ApplicationStatusHistoryEntry,
  Priority,
} from '../models/application.model';

@Injectable({ providedIn: 'root' })
export class ApplicationService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/applications`;

  list(params: {
    search?: string;
    status?: ApplicationStatus;
    priority?: Priority;
    companyId?: string;
    dateFrom?: string;
    dateTo?: string;
    page?: number;
    size?: number;
  } = {}) {
    let httpParams = new HttpParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        httpParams = httpParams.set(key, String(value));
      }
    });
    return this.http.get<Page<Application>>(this.baseUrl, { params: httpParams });
  }

  get(id: string) {
    return this.http.get<Application>(`${this.baseUrl}/${id}`);
  }

  getHistory(id: string) {
    return this.http.get<ApplicationStatusHistoryEntry[]>(`${this.baseUrl}/${id}/history`);
  }

  create(request: ApplicationRequest) {
    return this.http.post<Application>(this.baseUrl, request);
  }

  update(id: string, request: ApplicationRequest) {
    return this.http.put<Application>(`${this.baseUrl}/${id}`, request);
  }

  updateStatus(id: string, status: ApplicationStatus) {
    return this.http.patch<Application>(`${this.baseUrl}/${id}/status`, { status });
  }

  delete(id: string) {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
